File: GoogleNavigation/GoogleNavXCFrameworkDemos/Samples/RoutingOptionsViewController.m
/*
 * Copyright 2017 Google LLC. All rights reserved.
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this
 * file except in compliance with the License. You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
 * ANY KIND, either express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

#import "GoogleNavXCFrameworkDemos/Samples/RoutingOptionsViewController.h"

#import <CoreLocation/CoreLocation.h>

#if __has_feature(modules)
@import GoogleMaps;
#else
#import <GoogleMaps/GoogleMaps.h>
#endif
#if __has_feature(modules)
@import GoogleNavigation;
#else
#import <GoogleNavigation/GoogleNavigation.h>
#endif
#import "GoogleNavXCFrameworkDemos/Utils/NavDemoStringUtils.h"
#import "GoogleNavXCFrameworkDemos/Utils/NavDemoSwitch.h"
#import "GoogleNavXCFrameworkDemos/Utils/NavDemoUtilities.h"

static const double kMetersPerMile = 1609.34;

/* Accessibility identifier for the scroll view that contains the controls at the bottom. */
static NSString *const kControlsScrollViewIdentifier = @"ControlsScrollView";

/** The type of destination(s) that will be used to request the route. */
typedef NS_ENUM(NSInteger, DestinationType) {
  /** Custom destination waypoinds will be used.*/
  DestinationTypeCustom = 0,

  /** A single coordinate destination waypoint will be used. */
  DestinationTypeCoordinate,

  /** A single PlaceID destination waypoint will be used. */
  DestinationTypePlaceID,

  /** Multiple destination waypoints will be used. */
  DestinationTypeMultiple,
};

/** Helper function to deal with the nullable initializer of GMSNavigationWaypoint. */
static void AddWaypoint(NSMutableArray<GMSNavigationWaypoint *> *waypoints,
                        GMSNavigationWaypoint *_Nullable maybeWaypoint) {
  if (maybeWaypoint) {
    GMSNavigationWaypoint *waypoint = maybeWaypoint;
    [waypoints addObject:waypoint];
  }
}

/** Creates a test location-based waypoint. */
static GMSNavigationWaypoint *_Nullable CreateLocationWaypoint() {
  CLLocationCoordinate2D coordinate = CLLocationCoordinate2DMake(38.1013611111, -122.2551666667);
  NSString *title = @"United Financial Processing Group, Inc.";
  return [[GMSNavigationWaypoint alloc] initWithLocation:coordinate title:title];
}

/** Creates a test place-ID-based waypoint. */
static GMSNavigationWaypoint *_Nullable CreatePlaceWaypoint() {
  return [[GMSNavigationWaypoint alloc] initWithPlaceID:@"ChIJrXBqznMrm4ARMkLvKBKQYKQ"
                                                  title:@"Sacramento Airport"];
}

/** Returns a list of waypoints based on the currently selected destination type. */
static NSArray<GMSNavigationWaypoint *> *_Nullable CreateWaypoints(
    DestinationType destinationType) {
  NSMutableArray<GMSNavigationWaypoint *> *waypoints =
      [NSMutableArray<GMSNavigationWaypoint *> array];

  switch (destinationType) {
    case DestinationTypeCustom:
      return nil;
      break;
    case DestinationTypeCoordinate:
      AddWaypoint(waypoints, CreateLocationWaypoint());
      break;
    case DestinationTypePlaceID:
      AddWaypoint(waypoints, CreatePlaceWaypoint());
      break;
    case DestinationTypeMultiple:
      AddWaypoint(waypoints, CreateLocationWaypoint());
      AddWaypoint(waypoints, CreatePlaceWaypoint());
      break;
  }
  return waypoints;
}

@interface RoutingOptionsViewController () <GMSNavigatorListener,
                                            GMSRoadSnappedLocationProviderListener,
                                            GMSMapViewDelegate>

@end

@implementation RoutingOptionsViewController {
  DestinationType _destinationType;
  UILabel *_distanceTimeLabel;
  UILabel *_roadSnappedLocationLabel;
  UILabel *_routeInstructionLabel;
  UIButton *_continueToNextWaypointButton;
  NSMutableArray<GMSNavigationWaypoint *> *_waypoints;
  BOOL _isCustomRoutes;
  GMSNavigationMutableRoutingOptions *_routingOptions;
  UITextField *_targetDistanceTextField;
  GMSMapView *_mapView;
}

- (void)viewDidLoad {
  [super viewDidLoad];
  UIStackView *controls = self.controls;
  _mapView = [[GMSMapView alloc] init];
  _mapView.navigationEnabled = YES;
  _mapView.cameraMode = GMSNavigationCameraModeFollowing;
  _mapView.travelMode = GMSNavigationTravelModeDriving;
  [self.mainStackView addArrangedSubview:_mapView];
  _mapView.delegate = self;
  [_mapView.navigator addListener:self];
  _mapView.navigator.voiceGuidance = GMSNavigationVoiceGuidanceAlertsAndGuidance;
  _mapView.settings.compassButton = YES;
  _mapView.locationSimulator.speedMultiplier = 5.0f;
  [_mapView.roadSnappedLocationProvider addListener:self];
  [_mapView.roadSnappedLocationProvider startUpdatingLocation];
  _isCustomRoutes = YES;
  _routingOptions = [[GMSNavigationMutableRoutingOptions alloc]
      initWithRoutingStrategy:GMSNavigationRoutingStrategyDefaultBest];

  // Add a scroll view to contain the controls.
  UIScrollView *scrollView = [[UIScrollView alloc] init];
  scrollView.accessibilityIdentifier = kControlsScrollViewIdentifier;
  [self.mainStackView addArrangedSubview:scrollView];
  [scrollView.heightAnchor constraintEqualToConstant:240.f].active = YES;
  [scrollView addSubview:controls];

  // Constrain the width of the controls to the main view to enforce vertical scrolling.
  [controls.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor].active = YES;
  [controls.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor].active = YES;

  // Constrain the controls to be positioned vertically inside the scroll view.
  [controls.topAnchor constraintEqualToAnchor:scrollView.topAnchor].active = YES;
  [controls.bottomAnchor constraintEqualToAnchor:scrollView.bottomAnchor].active = YES;

  // Add each of the controls to the container view.
  [self addControls];

  // Simulate at a fixed location near Stanford University so the behaviour of this demo is
  // consistent.
  [_mapView.locationSimulator
      simulateLocationAtCoordinate:CLLocationCoordinate2DMake(37.436367, -122.167312)];
  _waypoints = [NSMutableArray array];
}

#pragma mark - Private

/** Creates the UI controls and adds them to the controls container. */
- (void)addControls {
  UIStackView *controls = self.controls;

  // Add buttons to request and clear routes.
  _routeInstructionLabel =
      GMSNavigationCreateLabelWithText(@"Tap the map to add a custom destination.");
  [controls addArrangedSubview:_routeInstructionLabel];
  UIStackView *routeControls = GMSNavigationCreateHorizontalStackView();
  [controls addArrangedSubview:routeControls];
  [routeControls addArrangedSubview:GMSNavigationCreateButton(self, @selector(requestRoute),
                                                              @"Request route")];
  [routeControls
      addArrangedSubview:GMSNavigationCreateButton(self, @selector(clearRoute), @"Clear route")];

  // Add buttons to start and stop guidance.
  UIStackView *guidanceControls = GMSNavigationCreateHorizontalStackView();
  [controls addArrangedSubview:guidanceControls];
  [guidanceControls addArrangedSubview:GMSNavigationCreateButton(self, @selector(startGuidance),
                                                                 @"Start guidance")];
  [guidanceControls addArrangedSubview:GMSNavigationCreateButton(self, @selector(stopGuidance),
                                                                 @"Stop guidance")];

  // Add button to go to next waypoint.
  _continueToNextWaypointButton = GMSNavigationCreateButton(
      self, @selector(continueToNextDestination), @"Continue to next waypoint");
  [_continueToNextWaypointButton setTitleColor:UIColor.whiteColor forState:UIControlStateNormal];
  [_continueToNextWaypointButton setTitleColor:UIColor.lightTextColor
                                      forState:UIControlStateDisabled];
  [controls addArrangedSubview:_continueToNextWaypointButton];

  // Add buttons to start and stop simulation of travel along the route.
  UIStackView *simulationControls = GMSNavigationCreateHorizontalStackView();
  [controls addArrangedSubview:simulationControls];
  [simulationControls
      addArrangedSubview:GMSNavigationCreateButton(self, @selector(simulateCurrentRoute),
                                                   @"Start simulating")];
  [simulationControls addArrangedSubview:GMSNavigationCreateButton(self, @selector(stopSimulation),
                                                                   @"Stop simulating")];

  // Add segmented controls to select the type of destinations.
  UILabel *destinationTypeLabel = GMSNavigationCreateLabelWithText(@"Destination Type");
  [controls addArrangedSubview:destinationTypeLabel];
  UISegmentedControl *destinationTypeSegmentedControl = GMSNavigationCreateSegmentedControl(
      self, @selector(updateDestinationType:), @[ @"Custom", @"Coordinate", @"PlaceID", @"Multi" ]);
  [controls addArrangedSubview:destinationTypeSegmentedControl];

  // Add segmented controls to select the travel mode.
  UILabel *travelModeLabel = GMSNavigationCreateLabelWithText(@"Travel Mode");
  [controls addArrangedSubview:travelModeLabel];
  UISegmentedControl *travelModeSegmentedControl = GMSNavigationCreateSegmentedControl(
      self, @selector(updateTravelMode:), @[ @"Driving", @"Cycling", @"Walking" ]);
  [controls addArrangedSubview:travelModeSegmentedControl];

  // Add an avoid highways switch.
  NavDemoSwitch *avoidHighwaysSwitch =
      [NavDemoSwitch switchWithLabel:@"Avoid highways"
                              target:self
                            selector:@selector(updateAvoidsHighways:)];
  [controls addArrangedSubview:avoidHighwaysSwitch];

  // Add an avoid tolls switch.
  NavDemoSwitch *avoidTollsSwitch = [NavDemoSwitch switchWithLabel:@"Avoid tolls"
                                                            target:self
                                                          selector:@selector(updateAvoidsTolls:)];
  [controls addArrangedSubview:avoidTollsSwitch];

  // Add segmented controls to select the routing strategy.
  UILabel *routingStrategyLabel = GMSNavigationCreateLabelWithText(@"Routing Strategy");
  [controls addArrangedSubview:routingStrategyLabel];
  UISegmentedControl *routingStrategySegmentedControl = GMSNavigationCreateSegmentedControl(
      self, @selector(updateRoutingStrategy:), @[ @"Default", @"Shorter", @"Target Distance" ]);
  [controls addArrangedSubview:routingStrategySegmentedControl];

  // Create text field for delta target distance.
  _targetDistanceTextField = [[UITextField alloc] init];
  _targetDistanceTextField.placeholder = @"Target Distance (mi)";
  _targetDistanceTextField.keyboardType = UIKeyboardTypeNumberPad;
  _targetDistanceTextField.hidden = YES;
  [_targetDistanceTextField createDoneCancelToolBar];
  [controls addArrangedSubview:_targetDistanceTextField];

  // Add segmented controls to select the alternate routes strategy.
  UILabel *alternateRoutesLabel = GMSNavigationCreateLabelWithText(@"Alternate Routes Strategy");
  [controls addArrangedSubview:alternateRoutesLabel];
  UISegmentedControl *alternateRoutesSegmentedControl = GMSNavigationCreateSegmentedControl(
      self, @selector(updateAlternateRoutesStrategy:), @[ @"All", @"None", @"One" ]);
  [controls addArrangedSubview:alternateRoutesSegmentedControl];

  // Add route and location information labels.
  _distanceTimeLabel = GMSNavigationCreateLabelWithText(@"");
  [self updateTimesAndDistances];
  [controls addArrangedSubview:_distanceTimeLabel];
  _roadSnappedLocationLabel = GMSNavigationCreateLabelWithText(@"ROAD-SNAPPED LOCATION\n");
  [controls addArrangedSubview:_roadSnappedLocationLabel];
}

/** Requests a route with the selected destination type, travel mode and options. */
- (void)requestRoute {
  if ((_destinationType != DestinationTypeCustom) && !_waypoints.count) {
    NSArray<GMSNavigationWaypoint *> *waypoints = CreateWaypoints(_destinationType);
    if (waypoints) {
      [_waypoints addObjectsFromArray:waypoints];
    }
  }
  [_mapView clear];
  [_mapView.navigator clearDestinations];
  if (_routingOptions.routingStrategy == GMSNavigationRoutingStrategyDeltaToTargetDistance &&
      !_targetDistanceTextField.isHidden) {
    double distance =
        ((_targetDistanceTextField.text ? [_targetDistanceTextField.text doubleValue] : 0) *
         kMetersPerMile);
    NSArray<NSNumber *> *temp = @[ @(distance) ];
    _routingOptions.targetDistancesMeters = temp;
  }
  __weak RoutingOptionsViewController *weakSelf = self;
  GMSRouteStatusCallback callback = ^(GMSRouteStatus routeStatus) {
    [weakSelf handleRouteCallbackWithStatus:routeStatus];
  };
  _continueToNextWaypointButton.enabled = NO;
  [_mapView.navigator setDestinations:_waypoints routingOptions:_routingOptions callback:callback];
}

/** Handles a route response with the given success or failure status. */
- (void)handleRouteCallbackWithStatus:(GMSRouteStatus)routeStatus {
  if (routeStatus == GMSRouteStatusOK) {
    _mapView.cameraMode = GMSNavigationCameraModeOverview;
  } else {
    // Show an error dialog to describe the failure.
    GMSNavigationPresentAlertController(self, GMSNavigationDemoMessageForRouteStatus(routeStatus),
                                        @"Route failed", @"OK");
  }
  _continueToNextWaypointButton.enabled = (_waypoints.count > 1);
}

/** Updates the destination type. */
- (void)updateDestinationType:(UISegmentedControl *)sender {
  _routeInstructionLabel.text = sender.selectedSegmentIndex == 0
                                    ? @"Tap the map to add a custom destination."
                                    : @"Tap \"Request Route\" for a preset destination.";
  _isCustomRoutes = sender.selectedSegmentIndex == 0 ? YES : NO;
  _destinationType = (DestinationType)(sender.selectedSegmentIndex);
  [_waypoints removeAllObjects];
}

/** Updates the travel mode. */
- (void)updateTravelMode:(UISegmentedControl *)sender {
  _mapView.travelMode = (GMSNavigationTravelMode)sender.selectedSegmentIndex;
}

/** Updates the routing strategy. */
- (void)updateRoutingStrategy:(UISegmentedControl *)sender {
  GMSNavigationRoutingStrategy newRoutingStrategy =
      (GMSNavigationRoutingStrategy)(sender.selectedSegmentIndex);
  if (_routingOptions.routingStrategy == GMSNavigationRoutingStrategyDeltaToTargetDistance ||
      newRoutingStrategy == GMSNavigationRoutingStrategyDeltaToTargetDistance) {
    [UIView animateWithDuration:0.3
                     animations:^{
                       _targetDistanceTextField.hidden =
                           newRoutingStrategy != GMSNavigationRoutingStrategyDeltaToTargetDistance;
                     }];
  }
  _routingOptions.routingStrategy = newRoutingStrategy;
}

/** Updates the alternate routes strategy. */
- (void)updateAlternateRoutesStrategy:(UISegmentedControl *)sender {
  _routingOptions.alternateRoutesStrategy =
      (GMSNavigationAlternateRoutesStrategy)sender.selectedSegmentIndex;
}

/** Updates the 'avoid highways' setting. */
- (void)updateAvoidsHighways:(NavDemoSwitch *)sender {
  _mapView.navigator.avoidsHighways = sender.on;
}

/** Updates the 'avoid tolls' setting. */
- (void)updateAvoidsTolls:(NavDemoSwitch *)sender {
  _mapView.navigator.avoidsTolls = sender.on;
}

/** Clears the current route if one is loaded. */
- (void)clearRoute {
  [_mapView.navigator clearDestinations];
  [_mapView clear];
  [_waypoints removeAllObjects];
}

/** Starts guidance. */
- (void)startGuidance {
  _mapView.navigator.guidanceActive = YES;
}

/** Stops guidance. */
- (void)stopGuidance {
  _mapView.navigator.guidanceActive = NO;
}

/** Continues to the next destination in a multi-waypoint route. */
- (void)continueToNextDestination {
  if (!_waypoints.count) {
    return;
  }
  [_waypoints removeObjectAtIndex:0];

  [self requestRoute];
}

/** Starts simulating along the current route. */
- (void)simulateCurrentRoute {
  [_mapView.locationSimulator simulateLocationsAlongExistingRoute];
}

/** Stops the simulation, returning the user location marker to the GPS-reported location. */
- (void)stopSimulation {
  [_mapView.locationSimulator stopSimulation];
}

/**
 * Formats the time and distance to each waypoint as text and displays it in the appropriate label.
 */
- (void)updateTimesAndDistances {
  NSMutableArray<NSString *> *timesAndDistances = [[NSMutableArray alloc] init];
  for (GMSNavigationWaypoint *waypoint in _waypoints) {
    NSTimeInterval time = [_mapView.navigator timeToWaypoint:waypoint];
    CLLocationDistance distance = [_mapView.navigator distanceToWaypoint:waypoint];
    if (time != CLTimeIntervalMax && distance != CLLocationDistanceMax) {
      NSDateComponentsFormatter *dateComponentsFormatter = [[NSDateComponentsFormatter alloc] init];
      NSString *timeString = [dateComponentsFormatter stringFromTimeInterval:time];
      [timesAndDistances
          addObject:[NSString stringWithFormat:@"%@\n  time:%@ dist:%.1fm", waypoint.title,
                                               timeString, distance]];
    } else {
      [timesAndDistances addObject:[NSString stringWithFormat:@"%@: unavailable.", waypoint.title]];
    }
  }
  NSString *timesAndDistancesText = [timesAndDistances componentsJoinedByString:@"\n"];
  _distanceTimeLabel.text =
      [NSString stringWithFormat:@"TIMES AND DISTANCES\n%@", timesAndDistancesText];
}

#pragma mark GMSMapViewDelegate

/**
 * Adds a marker at the coordinate tapped by the user. Locations are then added to the _waypoints
 * array. These waypoints will be used when creating routes.
 */
- (void)mapView:(GMSMapView *)mapView didTapAtCoordinate:(CLLocationCoordinate2D)coordinate {
  if (_isCustomRoutes) {
    GMSNavigationWaypoint *newWaypoint =
        [[GMSNavigationWaypoint alloc] initWithLocation:coordinate title:@"Custom Waypoint"];
    if (newWaypoint) {
      [_waypoints addObject:newWaypoint];
      GMSMarker *newWaypointMarker = [GMSMarker markerWithPosition:coordinate];
      newWaypointMarker.title = @"Custom Waypoint";
      newWaypointMarker.map = mapView;
    }
  }
}

#pragma mark GMSRoadSnappedLocationProviderListener

- (void)locationProvider:(GMSRoadSnappedLocationProvider *)locationProvider
       didUpdateLocation:(nonnull CLLocation *)location {
  // Format the road-snapped location as text and display it in the appropriate label.
  NSDateFormatter *dateFormat = [[NSDateFormatter alloc] init];
  dateFormat.dateFormat = @"YYYY-MM-dd\'  \'HH:mm:ss";
  NSString *dateString = [dateFormat stringFromDate:location.timestamp];
  NSString *locationText = [NSString
      stringWithFormat:
          @"ROAD-SNAPPED LOCATION\n"
          @"Lat: %f\rLng: %f\rAlt: %f\rhAcc: %f\rvAcc: %f\rCourse: %f\rSpeed: %f\rhTime: %@",
          location.coordinate.latitude, location.coordinate.longitude, location.altitude,
          location.horizontalAccuracy, location.verticalAccuracy, location.course, location.speed,
          dateString];
  _roadSnappedLocationLabel.text = locationText;
}

#pragma mark GMSNavigatorListener

- (void)navigator:(GMSNavigator *)navigator didUpdateRemainingTime:(NSTimeInterval)time {
  [self updateTimesAndDistances];
}

- (void)navigator:(GMSNavigator *)navigator
    didUpdateRemainingDistance:(CLLocationDistance)distance {
  [self updateTimesAndDistances];
}

@end

File: GoogleNavigation-Swift/GoogleNavSwiftXCFrameworkDemos/Swift/Samples/RoutingOptionsViewController.swift
/// Copyright 2020 Google LLC. All rights reserved.
///
///
/// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this
/// file except in compliance with the License. You may obtain a copy of the License at
///
///     http://www.apache.org/licenses/LICENSE-2.0
///
/// Unless required by applicable law or agreed to in writing, software distributed under
/// the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
/// ANY KIND, either express or implied. See the License for the specific language governing
/// permissions and limitations under the License.

import GoogleNavigation
import UIKit

class RoutingOptionsViewController: BaseSampleViewController {
  enum DestinationType: Int, CaseIterable, CustomStringConvertible {
    /// Use custom waypoints set by the user tapping on the map.
    case custom
    /// Use a waypoint created using coordinates.
    case coordinate
    /// Use a waypoints created using a place ID.
    case placeID
    /// Use multiple waypoints created using both coordinates and a place ID.
    case multi

    var description: String {
      switch self {
      case .custom:
        return "Custom"
      case .coordinate:
        return "Coordinate"
      case .placeID:
        return "PlaceID"
      case .multi:
        return "Multi"
      }
    }

    var instructionText: String {
      switch self {
      case .custom:
        return "Tap map to add waypoint."
      case .coordinate, .placeID, .multi:
        return #"Tap "Request route" for a preset destination."#
      }
    }
  }

  /// The main map view.
  private lazy var mapView: GMSMapView = {
    let mapView = GMSMapView()
    mapView.isNavigationEnabled = true
    mapView.settings.compassButton = true
    mapView.delegate = self
    return mapView
  }()

  /// Used to determine the type of waypoint destination(s) that will be used in this sample.
  private var destinationType: DestinationType = .custom {
    didSet {
      destinationTypeInstructionLabel.text = destinationType.instructionText
    }
  }

  /// The last type of waypoint destination(s) that was used in this sample.
  private var lastDestinationType: DestinationType = .custom

  /// Displays instructions for the current destination type.
  private lazy var destinationTypeInstructionLabel: UILabel = {
    let label = MenuUIHelpers.makeLabel(text: destinationType.instructionText)
    label.textAlignment = .center
    label.textColor = .systemRed
    return label
  }()

  /// Used to store custom waypoints set by the user.
  private var customWaypoints = [GMSNavigationWaypoint]()

  /// Returns the waypoints for the current destination type.
  private var currentWaypoints = [GMSNavigationWaypoint]()

  /// Receives a target distance from the user.
  private lazy var targetDistanceTextField: UITextField = {
    let textField = MenuUIHelpers.makeTextField(
      placeholder: "Target Distance (mi)",
      keyboardType: .numberPad)
    textField.addDoneCancelToolbar()
    return textField
  }()

  /// The routing options that will be passed to `setDestinations(_:routingOptions:callback:)`.
  private var routingOptions = GMSNavigationMutableRoutingOptions()

  override func viewDidLoad() {
    super.viewDidLoad()

    // Add a map.
    primaryStackView.insertArrangedSubview(mapView, at: 0)
    // Simulate at a fixed location near Stanford University.
    mapView.locationSimulator?.simulateLocation(
      at: CLLocationCoordinate2DMake(37.436367, -122.167312))

    addMenuSubview(destinationTypeInstructionLabel)

    // Add buttons to request/clear the route.
    addMenuSubview(
      MenuUIHelpers.makeStackView(arrangedSubviews: [
        MenuUIHelpers.makeMenuButton(
          title: "Request route",
          onTouchUpInside: (target: self, action: #selector(requestRoute))),
        MenuUIHelpers.makeMenuButton(
          title: "Clear route",
          onTouchUpInside: (target: self, action: #selector(clearRoute))),
      ]))

    // Add buttons to start/stop guidance.
    addMenuSubview(
      MenuUIHelpers.makeStackView(arrangedSubviews: [
        MenuUIHelpers.makeMenuButton(
          title: "Start guidance",
          onTouchUpInside: (target: self, action: #selector(startGuidance))),
        MenuUIHelpers.makeMenuButton(
          title: "Stop guidance",
          onTouchUpInside: (target: self, action: #selector(stopGuidance))),
      ]))

    // Add button to continue to the next waypoint.
    addMenuSubview(
      MenuUIHelpers.makeMenuButton(
        title: "Continue to next waypoint",
        onTouchUpInside: (target: self, action: #selector(continueToNextWaypoint))))

    // Add buttons to start/stop simulation.
    addMenuSubview(
      MenuUIHelpers.makeStackView(arrangedSubviews: [
        MenuUIHelpers.makeMenuButton(
          title: "Start simulation",
          onTouchUpInside: (target: self, action: #selector(startSimulation))),
        MenuUIHelpers.makeMenuButton(
          title: "Stop simulation",
          onTouchUpInside: (target: self, action: #selector(stopSimulation))),
      ]))

    // Add segmented controls to select the type of destinations.
    addMenuSubview(
      MenuUIHelpers.makeSegmentedControl(
        title: "Destination Type",
        segmentTitles: DestinationType.allCases.map(String.init(describing:)),
        onValueChanged: (target: self, action: #selector(destinationTypeControlDidUpdate))))

    // Add segmented controls to select the travel mode.
    addMenuSubview(
      MenuUIHelpers.makeSegmentedControl(
        title: "Travel Mode",
        segmentTitles: GMSNavigationTravelMode.allCases.map(String.init(describing:)),
        onValueChanged: (target: self, action: #selector(travelModeControlDidUpdate))))

    // Add an avoid highways switch.
    addMenuSubview(
      NavDemoSwitch(
        title: "Avoid highways",
        onValueChanged: (target: self, action: #selector(avoidHighwaysSwitchDidUpdate))))

    // Add an avoid tolls switch.
    addMenuSubview(
      NavDemoSwitch(
        title: "Avoid tolls",
        onValueChanged: (target: self, action: #selector(avoidTollsSwitchDidUpdate))))

    // Add segmented controls to select the routing strategy.
    addMenuSubview(
      MenuUIHelpers.makeSegmentedControl(
        title: "Routing Strategy",
        segmentTitles: GMSNavigationRoutingStrategy.allCases.map(String.init(describing:)),
        onValueChanged: (target: self, action: #selector(routingStrategyControlDidUpdate))))
    targetDistanceTextField.isHidden = true
    // Add a text field to get user input. This text field will only be visible if the routing
    // strategy is `.deltaToTargetDistance`.
    addMenuSubview(targetDistanceTextField)

    // Add segmented controls to select the alternate routes strategy.
    addMenuSubview(
      MenuUIHelpers.makeSegmentedControl(
        title: "Alternate Routes Strategy",
        segmentTitles: GMSNavigationAlternateRoutesStrategy.allCases.map(String.init(describing:)),
        onValueChanged: (target: self, action: #selector(alternateRoutesStrategyControlDidUpdate))))
  }

  private func createWaypoints() -> [GMSNavigationWaypoint] {
    switch destinationType {
    case .custom:
      return customWaypoints
    case .coordinate:
      if let ohloneCollege = GMSNavigationWaypoint(
        location: CLLocationCoordinate2DMake(37.529032, -121.918846),
        title: "Ohlone College")
      {
        return [ohloneCollege]
      }
    case .placeID:
      if let paramountTheatre = GMSNavigationWaypoint(
        placeID: "ChIJI2TVmK2Aj4ARxVVzT6uIQMw",
        title: "Paramount Theatre, Oakland")
      {
        return [paramountTheatre]
      }
    case .multi:
      if let ohloneCollege = GMSNavigationWaypoint(
        location: CLLocationCoordinate2DMake(37.529032, -121.918846),
        title: "Ohlone College"),
        let paramountTheatre = GMSNavigationWaypoint(
          placeID: "ChIJI2TVmK2Aj4ARxVVzT6uIQMw",
          title: "Paramount Theatre, Oakland")
      {
        return [ohloneCollege, paramountTheatre]
      }
    }
    return []
  }

  // MARK: - Menu handlers

  /// Requests a route with the selected destination type, travel mode and options.
  @objc private func requestRoute() {
    // If the destination type has changed, create new waypoints.
    if lastDestinationType != destinationType || currentWaypoints.isEmpty {
      currentWaypoints = createWaypoints()
      lastDestinationType = destinationType
    }

    // If the destination type is custom and new waypoints were created, update the custom waypoints
    if destinationType == .custom && !customWaypoints.isEmpty {
      currentWaypoints = customWaypoints
    }
    // Set the routing options target distance if `.deltaToTargetDistance` is the routing strategy.
    if routingOptions.routingStrategy == .deltaToTargetDistance
      && !targetDistanceTextField.isHidden,
      let value = Double(targetDistanceTextField.text ?? "")
    {
      let targetDistanceMiles = Measurement(value: value, unit: UnitLength.miles)
      let targetDistanceMeters = Int(targetDistanceMiles.converted(to: UnitLength.meters).value)
      // Convert the target distance from miles to meters.
      routingOptions.targetDistancesMeters = [targetDistanceMeters as NSNumber]
    }

    // Clear previous destinations and set the new destinations.
    let navigator = mapView.navigator
    clearRoute()
    navigator?.clearDestinations()
    navigator?.setDestinations(currentWaypoints, routingOptions: routingOptions) { routeStatus in
      self.handleRouteCallback(with: routeStatus)
    }
  }

  /// Clears the current route if one is loaded.
  @objc private func clearRoute() {
    mapView.navigator?.clearDestinations()
    if destinationType == .custom {
      customWaypoints.removeAll()
    }
    mapView.clear()
    mapView.cameraMode = .following
  }

  /// Starts guidance.
  @objc private func startGuidance() {
    mapView.cameraMode = .following
    mapView.navigator?.isGuidanceActive = true
  }

  /// Stops guidance.
  @objc private func stopGuidance() {
    mapView.navigator?.isGuidanceActive = false
  }

  /// Continues to the next destination in a multi-waypoint route.
  @objc private func continueToNextWaypoint() {
    if !currentWaypoints.isEmpty {
      currentWaypoints.removeFirst()
      if currentWaypoints.isEmpty {
        clearRoute()
      } else {
        requestRoute()
      }
    }
  }

  /// Starts simulating along the current route.
  @objc private func startSimulation() {
    mapView.locationSimulator?.simulateLocationsAlongExistingRoute()
  }

  /// Stops the simulation, returning the user location marker to the GPS-reported location.
  @objc private func stopSimulation() {
    mapView.locationSimulator?.stopSimulation()
  }

  /// Updates the destination type.
  @objc private func destinationTypeControlDidUpdate(_ segmentedControl: UISegmentedControl) {
    destinationType = DestinationType(rawValue: segmentedControl.selectedSegmentIndex) ?? .custom
  }

  /// Updates the travel mode.
  @objc private func travelModeControlDidUpdate(_ segmentedControl: UISegmentedControl) {
    mapView.travelMode =
      GMSNavigationTravelMode(rawValue: segmentedControl.selectedSegmentIndex) ?? .driving
  }

  /// Updates the 'avoid highways' setting.
  @objc private func avoidHighwaysSwitchDidUpdate(_ sender: UISwitch) {
    mapView.navigator?.avoidsHighways = sender.isOn
  }

  /// Updates the 'avoid tolls' setting.
  @objc private func avoidTollsSwitchDidUpdate(_ sender: UISwitch) {
    mapView.navigator?.avoidsTolls = sender.isOn
  }

  /// Updates the routing strategy.
  @objc private func routingStrategyControlDidUpdate(_ segmentedControl: UISegmentedControl) {
    let newRoutingStrategy =
      GMSNavigationRoutingStrategy(rawValue: segmentedControl.selectedSegmentIndex) ?? .defaultBest

    // Show/hide the target distance textField as appropriate.
    if routingOptions.routingStrategy == .deltaToTargetDistance
      || newRoutingStrategy == .deltaToTargetDistance
    {
      UIView.animate(withDuration: BaseSampleViewController.menuAnimationsDuration) {
        self.targetDistanceTextField.isHidden = (newRoutingStrategy != .deltaToTargetDistance)
      }
    }

    routingOptions.routingStrategy = newRoutingStrategy
  }

  /// Updates the alternate routes strategy.
  @objc private func alternateRoutesStrategyControlDidUpdate(
    _ segmentedControl: UISegmentedControl
  ) {
    routingOptions.alternateRoutesStrategy =
      GMSNavigationAlternateRoutesStrategy(rawValue: segmentedControl.selectedSegmentIndex) ?? .all
  }

  // MARK: - Private helpers

  /// Handles a route response with the given success or failure status.
  private func handleRouteCallback(with routeStatus: GMSRouteStatus) {
    if routeStatus == .OK {
      mapView.cameraMode = .overview
    } else {
      // Show an error dialog to describe the failure.
      let message = routeStatus.description
      showErrorMessage(title: "Route failed", message: message)
    }
  }

  /// Shows an error alert.
  ///
  /// - Note: This is not normally the best way to handle problems, but is used here for
  /// demonstration purposes.
  private func showErrorMessage(title: String, message: String?) {
    let alertController = UIAlertController(title: title, message: message, preferredStyle: .alert)
    alertController.addAction(UIAlertAction(title: "OK", style: .default))
    present(alertController, animated: true)
  }
}

// MARK: - Extensions

extension RoutingOptionsViewController: GMSMapViewDelegate {
  func mapView(_ mapView: GMSMapView, didTapAt coordinate: CLLocationCoordinate2D) {
    guard destinationType == .custom else { return }

    // Create a waypoint at the location that was tapped on.
    let waypointTitle = "Custom waypoint \(customWaypoints.count)"
    guard
      let newWaypoint = GMSNavigationWaypoint(
        location: coordinate,
        title: waypointTitle)
    else {
      showErrorMessage(title: "Failed to create custom waypoint", message: nil)
      return
    }

    // Store the waypoint to use later when calling setDestinations.
    customWaypoints.append(newWaypoint)

    // Add a custom marker to make it easier to track where they've been placed. This is not a
    // necessary step and is just for convenience.
    let marker = GMSMarker(position: coordinate)
    marker.title = waypointTitle
    marker.map = mapView
  }
}

extension GMSNavigationTravelMode: CaseIterable, CustomStringConvertible {
  public static var allCases: [GMSNavigationTravelMode] {
    return [.driving, .cycling, .walking, .twoWheeler, .taxicab]
  }

  public var description: String {
    switch self {
    case .driving:
      return "Driving"
    case .cycling:
      return "Cycling"
    case .walking:
      return "Walking"
    case .twoWheeler:
      return "Two Wheeler"
    case .taxicab:
      return "Taxicab"
    @unknown default:
      fatalError()
    }
  }
}

extension GMSNavigationAlternateRoutesStrategy: CaseIterable, CustomStringConvertible {
  public static var allCases: [GMSNavigationAlternateRoutesStrategy] {
    return [.all, .none, .one]
  }

  public var description: String {
    switch self {
    case .all:
      return "All"
    case .none:
      return "None"
    case .one:
      return "One"
    @unknown default:
      fatalError()
    }
  }
}

extension GMSNavigationRoutingStrategy: CaseIterable, CustomStringConvertible {
  public static var allCases: [GMSNavigationRoutingStrategy] {
    return [.defaultBest, .shorter, .deltaToTargetDistance]
  }

  public var description: String {
    switch self {
    case .defaultBest:
      return "Default"
    case .shorter:
      return "Shorter"
    case .deltaToTargetDistance:
      return "Target Distance"
    @unknown default:
      fatalError()
    }
  }
}
